Sources/Plasma/CoreLib/hsSystemInfo.cpp

```diff
@@ -0,0 +1,300 @@
+                    }
+                }
+            }
+            s.Close();
+        }
+    }
+
+    return system;
+}
+
+ST::string hsSystemInfo::GetOperatingSystem()
+{
+#ifdef HS_BUILD_FOR_WIN32
+    ST::string system;
+    const RTL_OSVERSIONINFOEXW& info = hsGetWindowsVersion();
+
+    typedef const char* (CDECL * wine_get_version)();
+    typedef void (CDECL * wine_get_host_version)(const char** sysname, const char** release);
+
+    HMODULE ntdll = GetModuleHandleW(L"ntdll");
+    if (ntdll) {
+        wine_get_version wine_target_version = (wine_get_version)GetProcAddress(ntdll, "wine_get_version");
+        wine_get_host_version wine_host_version = (wine_get_host_version)GetProcAddress(ntdll, "wine_get_host_version");
+        if (wine_target_version && wine_host_version) {
+            // WINE for some reason likes to pretend to be Windows Server. Debunk that crap
+            // because we really only want to get "Windows Server 2008 R2" if they are really
+            // on that OS, not because WINE is saying it's a Windows 7 server. Ugh.
+            ST::string windowsVersion = IGetWindowsVersion(info, false);
+
+            // Ideally, we would try to parse Z:\etc\os-release to get the Linux distro name,
+            // but that seems to crash WINE, unfortunately. So, we'll just naively use whatever
+            // WINE returns (the kernel name and version).
+            const char* sysname;
+            const char* release;
+            wine_host_version(&sysname, &release);
+            const char* wineVersion = wine_target_version();
+
+            system = ST::format("{} [Emulated on {} ({}) by WINE {}]", windowsVersion, sysname, release, wineVersion);
+        }
+    }
+    if (system.empty())
+        system = IGetWindowsVersion(info);
+
+    return system;
+#else
+    ST::string system;
+
+    // Try linux first so we can get the distro name.
+    system = IGetLinuxVersion("/etc/os-release");
+    if (!system.empty())
+        return system;
+
+    // Should work for everything else.
+    utsname sysinfo;
+    uname(&sysinfo);
+    system = ST::format("{} {} ({})", sysinfo.sysname, sysinfo.release, sysinfo.machine);
+
+    return system;
+#endif
+}
+
+uint64_t hsSystemInfo::GetRAM()
+{
+#if defined(HS_BUILD_FOR_WIN32)
+    ULONGLONG memoryInKB = 0;
+    GetPhysicallyInstalledSystemMemory(&memoryInKB);
+    hsAssert((memoryInKB % 1024) == 0, "Odd amount of memory?");
+    return (uint64_t)(memoryInKB / 1024ULL);
+#elif defined(HAVE_SYSINFO)
+    struct sysinfo info;
+    if (sysinfo(&info) == 0) {
+        uint64_t desiredUnit = 1024 * 1024;
+        if (info.mem_unit != desiredUnit)
+            return ((uint64_t)info.totalram * (uint64_t)info.mem_unit) / desiredUnit;
+        return info.totalram;
+    }
+#elif defined(HAVE_SYSCTL)
+    int mib[2]{ CTL_HW, HW_MEMSIZE };
+    uint64_t memory = 0;
+    size_t length = sizeof(memory);
+    if (sysctl(mib, sizeof(mib), &memory, &length, nullptr, 0) != 0)
```

Comment:
I think the second argument to `sysctl` is supposed to be the length of the array, not the size of the array